src/checkqueue.h

```diff
@@ -181,24 +174,16 @@ class CCheckQueue
         }
     }
 
-    //! Stop all of the worker threads.
-    void StopWorkerThreads() EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+    ~CCheckQueue()
     {
-        WITH_LOCK(m_mutex, m_request_stop = true);
+        m_request_stop = true;
```

Comment:
I think that removing the lock here could cause a deadlock.
`m_request_stop` is set to true and then `notify_all()` is called; but when there is a worker thread currently standing inside the `while (queue.empty() && !m_request_stop) {` loop but before the `cond.wait(lock);` it will not receive the notification and then proceed to wait in the `cond` indefinitely.

That deadlock was not possible before when the `m_mutex` had to be acquired to change `m_request_stop`, because then it was guaranteed that all threads were either waiting on the condition variable or waiting outside that loop to get the lock

Reply:
Doesn't the default `memory_order_seq_cst` memory order for `std::atomic<bool> m_request_stop` guarantee the correct behaviour?

Reply:
The only thing guaranteed is the order; so `m_request_stop` will be set to `true` before `notify_all` is called. But there is no guarantee where the other threads currently are because there is now no lock anymore. E.g. a worker thread could be currently here:

```cpp
while (queue.empty() && !m_request_stop) {
    if (fMaster && nTodo == 0) {   // <-------- Worker thread waiting here due to context switch
        nTotal--;
        bool fRet = fAllOk;
        // reset the status for new work later
        fAllOk = true;
        // return the current status
        return fRet;
    }
    nIdle++;
    cond.wait(lock); // wait
    nIdle--;
}
```

Now while the worker thread is not running for whatever reason (context switch, high CPU load, ...) the main thread runs, calls `m_request_stop = true`, then `m_worker_cv.notify_all();`, and then hangs in the `t.join()` loop.

Now for whatever reason the worker thread get's its share of the CPU and continues running, and will run right into the `cond.wait(lock);` without checking `m_request_stop`. There it won't receive the notification because it came before it called the `wait()`, so the worker thread will now wait.

So now the master hangs in the `join()` loop, and the worker waits for notification => deadlock.

Reply:
You're right. Going to amend the patch.

Reply:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/26762#issuecomment-1500980698).